README
Net::Radius Modules

The modules included here provide an interface to the RADIUS
protocol. It consists of the following modules:

Net::Radius::Packet	- Deals with RADIUS packets
Net::Radius::Dictionary	- Deals with RADIUS dictionaries

This module is essentially the original RADIUS-1.0 distribution by
Christopher Masto plus a number of changes and fixes by Luis Munoz and
Ian Smith.

It has been changed so that it better fixes the CPAN namespace. See
the other README.* files in this archive for additional information.

The installation follows the standard protocol...

$ perl Makefile.PL
$ make
$ make test
$ make install

The ./examples directory contain a number of simple examples.

This code supports the use of vendor specific attributes. This
type of attribute is defined in RFC-2138 and is used to support
'propietary' extensions on top of the base RADIUS specification.

There are two new kinds of entries in the RADIUS dictionary in
order to specify VSAs.

VENDORATTR <vendor> <attribute> <id> <type>

This entry is used to create a new kind of vendor attribute,
such as in this example

VENDORATTR 	9 	cisco-avpair 	1 	string

This creates a new vendor-specific attribute for vendor 9 (Cisco
Systems), with name 'cisco-avpair'. This attribute is identified by
numeric id '1' and is associated with a string value.

The second type of entry allows the specification of named values.
The following is an hypotetical example of named value entry.

VENDORATTR	9	cisco-enum	254	integer
VENDORVALUE	9	cisco-enum	Value-1	1
VENDORVALUE	9	cisco-enum	Value-2	2
VENDORVALUE	9	cisco-enum	Value-3	3

About the stability, this code has been in very active use at a
largish ISP with hundreds of thousands of users using a variety of
network equipment with impressive results. It has been succesfully
used under FreeBSD, Linux, Solaris and Tru64.

A sample dictionary file is included, mostly to ilustrate the syntax
of the expected file.

Enjoy...



Changes
Revision history for Perl extension RADIUS.

1.55	Thu Apr 26 16:21:46 VET 2007
	* Darrian Hale reported that $p->unset_attr('Attribute',
	  $packet->attr('Attribute')) was b0rken due to the attribute
	  lookup using the "escaped" value instead of the raw value
	  returned by ->attr. Escaping was removed and a test was
	  added

	* Versions bumped to 1.55 as prior fixes were confirmed to
          work - Release to CPAN

	Sun Apr 22 16:54:03 VET 2007
	* Fixed the VSA packer to accept a vendor-id and a vendor
	  name. This will help if similar bugs are still lurking
	  around. This was pointed to by Pedro as well

	Sat Apr 21 14:04:35 VET 2007
	* Fixed VSA packing bug pointed out by Pedro Cavaca. Parts of
	  Packet.pm expected to receive a numeric vendor id, others
	  expected the string name


1.54	Tue Jan 30 11:13:42 VET 2007
	* Added missing packet file to MANIFEST
	* Fixed an innocuous warn() during make test

1.53	Tue Jan 30 06:25:33 VET 2007
	* Packets with VSAs that couldn't be decoded caused an endless
	  loop. This has now been fixed by skipping over the b0rked
	  attribute, thus parsing as much of the packet as possible. A
	  warn() is issued
	* The packet tests now fail if the decode causes a warn() -
	  This may be helpful in the future for catching subtle
	  errors, specially in the decoding of packets
	* Added a new test packet for Cisco VPN 3000 - Thanks Luis

1.52    Mon Jan  8 15:47:38 VET 2007
	* Added patch by Freman (consistend ->set_vsattr and
	  ->unset_vsattr) - Thanks Freman
	* ->attr_slot deprecated. ->attr_slot_name and ->attr_slot_val
	  introduced
	* Added documentation for those methods
	* Added tests to verify correct operation of the overwrite and
	  slots functionality
	* Multiple attributes (ie, multiple Proxy-State attributes)
	  are now correctly handled
	* Include (untested) patch from j7 (Thanks) which allows for
	  multiple VSAs. We need some testing for this...

	Tue Jan  9 15:52:06 VET 2007
	* Added support for the packet decoding tests as well as
	  samples of common packets in a live production environment
	* dictionary.base now replaces dictionary.orig

	Sun Jan 14 14:54:25 VET 2007
	* Added BroadSoft VSA dictionary supplied by j7 - Fixed typo
	within it
	* Extracted binary packet from the sample provided by j7 and
	added standard packet decoding test	

1.51	Mon Nov 13 21:08:28 VET 2006
	* Improve the ->dump() and ->str_dump() in ::Packet
	* Add the ->auth_*_verify() methods and tests
	* Dictionaries can now use the vendor name everywhere
	* Added all the dictionaries from Ethereal for completeness
	* Shipped dictionaries were cleaned up. IMPORTANT: Legacy,
	  obsolete entries have been commented. Please review the
	  dictionary files or keep the ones currently in production
	  before installing.
	* Added pseudo-support for ipv6addr, ifid and date types
	* Limit the packet authenticator to 16 bytes
	* Tagged attribue support is broken/incomplete - Need examples
	  to properly write tests and fill in missing functionality
	* ->password() now accepts an optional attribute to work with
	  instead of the default 'User-Password'

1.50	Mon Oct 23 15:43:20 VET 2006
	Applied patch by Chris that adds support for tagged
	attributes.
	Applied patch from Flavio, adding this:
	* [Dictionary.pm] accept a new keyword in dictionary file, namely
	  "PACKET", that allows to define packet types
	* [Dictionary.pm] these packet types default to the current
	  state-of-the-art according to 
	  http://www.iana.org/assignments/radius-types
	* [Dictionary.pm] the packet types can be accessed via ad-hoc
          methods
	* [Dictionary.pm] the constructor has been extended to accept
	  a list of dictionary files to load upon creation (instead of
	  a single one)
	* [Packet.pm] hardcoded mapping hashes in pack() and unpack()
	  methods are now loaded from the dictionary
	Bumped version to 1.50 in the .pm files.

1.49	Wed Aug  9 11:57:52 VET 2006
	- Making dictionaries case-insensitive seems to breaks old
	code in hard to debug ways. This decision was officially
	reversed
	- set_password now accepts an optional attribute argmuent, to
	tell the code where to store the encripted password. Still
	defaults to 'User-Password'

1.46	Mon Jul 31 17:28:16 VET 2006
	- Dictionaries now can be merged with each other, as suggested
	by Kevin Steves
	- Accessors for dictionary entries now use lc() to match what
	is done at parse time
	- Dictionaries are now officially case-insensitive
	- Various documentation improvements and updates
	- Added code against bug in Alcatel 5620 SAM Release 3.0
	(Improper attribute type). Possibly helpful in other cases as
	well

1.45	Thu Oct 16 10:45:02 VET 2003
	- Added support for "octets" as requested by Alex Chen
	- Improved / added tests. This now requires Test::More
	- Understands the VENDOR syntax for FreeRadius

1.44	Tue Jan  7 23:06:12 VET 2003
	- By default, tuples not in the dictionary are now ignored.
	- Added ->show_unknown_entries() to issue warn()s for each unknown
	  tuple found in the packets.

1.43	Thu Dec 20 15:36:42 GMT 2001
	Changes by Tony Mountifield <tony@mountifield.org>
	- added set_password method to generate client password correctly
	- added unset_attr method to remove an attribute from a Packet
	- added example-client.pl to demonstrate auth and acct
	- Added User-Password (as per RFC 2138) both in the code and
	in the dictionaries [There's some (broken) software that insists
	in 'Password', so this is used unless 'User-Password' is present -lem]
	- corrected syntax of the "Welcome, Larry" reply message example.

1.42    Fri Jul  6 12:22:36 VET 2001
	- Fixed sobe warnings in the packet dump code. (Thanks to Kevin Payne).

1.41	Fri May 11 10:54:19 EDT 2001
	Changes by Jim Harle <harle@usna.edu> (Thanks Jim! -lem)
	- added 3com dictionary
	- added example radius<->ldap server
	- fixed examples to show correct module name
	- made Digest::MD5 a prerequisite in Makefile.PL
	Dictionary.pm changes

Radius/Dictionary.pm
package Net::Radius::Dictionary;

use strict;
use warnings;
use vars qw($VERSION);

# $Id: Dictionary.pm,v 1.11 2007/04/26 20:20:02 lem Exp $

$VERSION = '1.55';

sub new {
  my $class = shift;
  my $self = { 
      rvsattr	=> {},
      vsattr	=> {},
      vsaval	=> {},
      rvsaval	=> {},
      attr	=> {},
      rattr	=> {},
      val	=> {},
      rval	=> {},
      vendors	=> {},
      packet    => undef,  # Fall back to default
      rpacket   => undef,  # Fall back to default
  };
  bless $self, $class;
  $self->readfile($_) for @_; # Read all given dictionaries
  return $self;
}

sub readfile {
  my ($self, $filename) = @_;

  open DICT, "<$filename";

  while (defined(my $l = <DICT>)) {
    next if $l =~ /^\#/;
    next unless my @l = split /\s+/, $l;

    if ($l[0] =~ m/^vendor$/i) 
    {
	if (defined $l[1] and defined $l[2] and $l[2] =~ /^[xo0-9]+$/)
	{
	    if (substr($l[2],0,1) eq "0") { #allow hex or octal
		my $num = lc($l[2]);
		$num =~ s/^0b//;
		$l[2] = oct($num);
	    }   
	    $self->{vendors}->{$l[1]} = $l[2];
	}
	else
	{
	    warn "Garbled VENDOR line $l\n";
	}
    }
    elsif ($l[0] =~ m/^attribute$/i) 
    {
	if (@l == 4)
	{
	    $self->{attr}->{$l[1]}  = [@l[2,3]];
	    $self->{rattr}->{$l[2]} = [@l[1,3]];
	}
	elsif (@l == 5)		# VENDORATTR
	{
	    if (substr($l[2],0,1) eq "0") { #allow hex or octal
		my $num = lc($l[2]);
		$num =~ s/^0b//;
		$l[2] = oct($num);
	    }   
	    if (exists $self->{vendors}->{$l[4]})
	    {
		$self->{vsattr}->{$self->{vendors}->{$l[4]}}->{$l[1]} 
		= [@l[2, 3]];
		$self->{rvsattr}->{$self->{vendors}->{$l[4]}}->{$l[2]} 
		= [@l[1, 3]];
	    }
	    elsif ($l[4] =~ m/^\d+$/)
	    {
		$self->{vsattr}->{$l[4]}->{$l[1]}	= [@l[2, 3]];
		$self->{rvsattr}->{$l[4]}->{$l[2]}	= [@l[1, 3]];
	    }
	    else
	    {
		warn "Warning: Unknown vendor $l[4]\n";
	    }
	}
    }
    elsif ($l[0] =~ m/^value$/i) {
      if (exists $self->{attr}->{$l[1]}) {
	  $self->{val}->{$self->{attr}->{$l[1]}->[0]}->{$l[2]}  = $l[3];
	  $self->{rval}->{$self->{attr}->{$l[1]}->[0]}->{$l[3]} = $l[2];
      }
      else {
	  for my $v (keys %{$self->{vsattr}})
	  {
	      if (defined $self->{vsattr}->{$v}->{$l[1]})
	      {
		  $self->{vsaval}->{$v}->{$self->{vsattr}->{$v}
					      ->{$l[1]}->[0]}->{$l[2]} 
		  = $l[3];
		  $self->{rvsaval}->{$v}->{$self->{vsattr}->{$v}
					   ->{$l[1]}->[0]}->{$l[3]} 
		  = $l[2];
	      }
	  }
      }
    }
    elsif ($l[0] =~ m/^vendorattr$/i) {
	if (substr($l[3],0,1) eq "0") { #allow hex or octal
          my $num = lc($l[3]);
          $num =~ s/^0b//;
          $l[3] = oct($num);
	}   
	if (exists $self->{vendors}->{$l[1]})
	{
	    $self->{vsattr}->{$self->{vendors}->{$l[1]}}->{$l[2]} 
	    = [@l[3, 4]];
	    $self->{rvsattr}->{$self->{vendors}->{$l[1]}}->{$l[3]} 
	    = [@l[2, 4]];
	}
	elsif ($l[1] =~ m/^\d+$/)
	{
	    $self->{vsattr}->{$l[1]}->{$l[2]} = [@l[3, 4]];
	    $self->{rvsattr}->{$l[1]}->{$l[3]} = [@l[2, 4]];
	}
	else
	{
	    warn "Warning: Unknown vendor $l[1]\n";
	}
    }
    elsif ($l[0] =~ m/^vendorvalue$/i) {
	if (substr($l[4],0,1) eq "0") 
	{ #allow hex or octal 
          my $num = lc($l[4]);
          $num =~ s/^0b//;
          $l[4] = oct($num);
	}
	if (exists $self->{vendors}->{$l[1]})
	{
	    $self->{vsaval}->{$self->{vendors}->{$l[1]}}
	    ->{$self->{vsattr}->{$self->{vendors}->{$l[1]}}
	       ->{$l[2]}->[0]}->{$l[3]} = $l[4];
	    $self->{rvsaval}->{$self->{vendors}->{$l[1]}}
	    ->{$self->{vsattr}->{$self->{vendors}->{$l[1]}}
	       ->{$l[2]}->[0]}->{$l[4]} = $l[3];
	}
	elsif ($l[1] =~ m/^\d+$/)
	{
	    $self->{vsaval}->{$l[1]}->{$self->{vsattr}->{$l[1]}->{$l[2]}
				       ->[0]}->{$l[3]} = $l[4];
	    $self->{rvsaval}->{$l[1]}->{$self->{vsattr}->{$l[1]}->{$l[2]}
					->[0]}->{$l[4]} = $l[3];
	}
	else {
	    warn "Warning: $filename contains vendor value for ",
	    "unknown vendor attribute - ignored ",
	    "\"$l[1]\"\n  $l";
	}
    }
    elsif (lc($l[0]) eq 'packet') {
        my ($name, $value) = @l[1,2];
        $self->{packet}{$name} = $value;
        $self->{rpacket}{$value} = $name;
    }
    else {
      warn "Warning: Weird dictionary line: $l\n";
    }
  }
  close DICT;
}

# Accessors for standard attributes

sub vendor_num	 { $_[0]->{vendors}->{$_[1]};		}
sub attr_num     { $_[0]->{attr}->{$_[1]}->[0];		}
sub attr_type    { $_[0]->{attr}->{$_[1]}->[1];		}
sub attr_name    { $_[0]->{rattr}->{$_[1]}->[0];	}
sub attr_numtype { $_[0]->{rattr}->{$_[1]}->[1];	}
sub attr_has_val { $_[0]->{val}->{$_[1]};		}
sub val_has_name { $_[0]->{rval}->{$_[1]};		}
sub val_num      { $_[0]->{val}->{$_[1]}->{$_[2]};	}
sub val_name     { $_[0]->{rval}->{$_[1]}->{$_[2]};	}
sub val_tag      { $_[0]->{val}->{$_[1]}->{$_[3]}; 	}

# Accessors for Vendor-Specific Attributes

sub vsattr_num      { $_[0]->{vsattr}->{$_[1]}->{$_[2]}->[0];		}
sub vsattr_type     { $_[0]->{vsattr}->{$_[1]}->{$_[2]}->[1];		}
sub vsattr_name     { $_[0]->{rvsattr}->{$_[1]}->{$_[2]}->[0];		}
sub vsattr_numtype  { $_[0]->{rvsattr}->{$_[1]}->{$_[2]}->[1];		}
sub vsattr_has_val  { $_[0]->{vsaval}->{$_[1]}->{$_[2]};		}
sub vsaval_has_name { $_[0]->{rvsaval}->{$_[1]}->{$_[2]};		}
sub vsaval_has_tval { $_[0]->{vsaval}->{$_[1]}->{$_[2]}->[0];		}
sub vsaval_has_tag  { $_[0]->{vsaval}->{$_[1]}->{$_[2]}->[1];		}
sub vsaval_num      { $_[0]->{vsaval}->{$_[1]}->{$_[2]}->{$_[3]};	}
sub vsaval_name     { $_[0]->{rvsaval}->{$_[1]}->{$_[2]}->{$_[3]};	}

# Accessors for packet types. Fall-back to defaults if the case.

# Defaults taken from http://www.iana.org/assignments/radius-types
# as of Oct 21, 2006
my %default_packets = (
    'Access-Request'      => 1, # [RFC2865]
    'Access-Accept'       => 2, # [RFC2865]
    'Access-Reject'       => 3, # [RFC2865]
    'Accounting-Request'  => 4, # [RFC2865]
    'Accounting-Response' => 5, # [RFC2865]
    'Accounting-Status'   => 6, # [RFC2882] (now Interim Accounting)
    'Interim-Accounting'  => 6, # see previous note
    'Password-Request'    => 7, # [RFC2882]
    'Password-Ack'        => 8, # [RFC2882]
    'Password-Reject'     => 9, # [RFC2882]
    'Accounting-Message'  => 10, # [RFC2882]
    'Access-Challenge'    => 11, # [RFC2865]
    'Status-Server'       => 12, # (experimental) [RFC2865]
    'Status-Client'       => 13, # (experimental) [RFC2865]
    'Resource-Free-Request'   => 21, # [RFC2882]
    'Resource-Free-Response'  => 22, # [RFC2882]
    'Resource-Query-Request'  => 23, # [RFC2882]
    'Resource-Query-Response' => 24, # [RFC2882]
    'Alternate-Resource-Reclaim-Request' => 25, # [RFC2882]
    'NAS-Reboot-Request'  => 26, # [RFC2882]
    'NAS-Reboot-Response' => 27, # [RFC2882]
    # 28       Reserved
    'Next-Passcode'       => 29, # [RFC2882]
    'New-Pin'             => 30, # [RFC2882]
    'Terminate-Session'   => 31, # [RFC2882]
    'Password-Expired'    => 32, # [RFC2882]
    'Event-Request'       => 33, # [RFC2882]
    'Event-Response'      => 34, # [RFC2882]
    'Disconnect-Request'  => 40, # [RFC3575]
    'Disconnect-ACK'      => 41, # [RFC3575]
    'Disconnect-NAK'      => 42, # [RFC3575]
    'CoA-Request'         => 43, # [RFC3575]
    'CoA-ACK'             => 44, # [RFC3575]
    'CoA-NAK'             => 45, # [RFC3575]
    'IP-Address-Allocate' => 50, # [RFC2882]
    'IP-Address-Release'  => 51, # [RFC2882]
    # 250-253  Experimental Use
    # 254      Reserved
    # 255      Reserved  [RFC2865]
);

# Reverse defaults. Remember that code #6 has a double mapping, force
# to Interim-Accouting
my %default_rpackets 
    = map { $default_packets{$_} => $_ } keys %default_packets;
$default_rpackets{6} = 'Interim-Accounting';

# Get full hashes
sub packet_numbers { %{ $_[0]->{packet}  || \%default_packets  }  }
sub packet_names   { %{ $_[0]->{rpacket} || \%default_rpackets }; }

# Single resolution, I'm taking care of avoiding auto-vivification
sub packet_hasname {
    my $href = $_[0]->{packet} || \%default_packets;
    my $ok = exists $href->{$_[1]};
    return $ok unless wantarray;
    # return both answer and the resolution
    return ($ok, $ok ? $href->{$_[1]} : undef);
}

sub packet_hasnum {
    my $href = $_[0]->{rpacket} || \%default_rpackets;
    my $ok = exists $href->{$_[1]};
    return $ok unless wantarray;
    # return both answer and the resolution
    return ($ok, $ok ? $href->{$_[1]} : undef);
}

# Note: crossed, as it might not be immediately evident
sub packet_num  { ($_[0]->packet_hasname($_[1]))[1]; }
sub packet_name { ($_[0]->packet_hasnum($_[1]))[1];  }

1;
__END__

=head1 NAME

Net::Radius::Dictionary - RADIUS dictionary parser

=head1 SYNOPSIS

  use Net::Radius::Dictionary;

  my $dict = new Net::Radius::Dictionary "/etc/radius/dictionary";
  $dict->readfile("/some/other/file");
  my $num = $dict->attr_num('User-Name');
  my $name = $dict->attr_name(1);
  my $vsa_num = $dict->vsattr_num(9, 'cisco-avpair');
  my $vsa_name = $dict->vsattr_name(9, 1);

=head1 DESCRIPTION

This is a simple module that reads a RADIUS dictionary file and
parses it, allowing conversion between dictionary names and numbers.
Vendor-Specific attributes are supported in a way consistent to the
standards.

A few earlier versions of this module attempted to make dictionaries
case-insensitive. This proved to be a very bad decision. From this
version on, this tendency is reverted: Dictionaries and its contents
are to be case-sensitive to prevent random, hard to debug failures in
production code.

=head2 METHODS

=over

=item B<new($dict_file, ...)>

Returns a new instance of a Net::Radius::Dictionary object. This
object will have no attributes defined, as expected.

If given an (optional) list of filenames, it calls I<readfile> for you
for all of them, in the given order.

=item B<-E<gt>readfile($dict_file)>

Parses a dictionary file and learns the mappings to use. It can be
called multiple times for the same object. The result will be that new
entries will override older ones, thus you could load a default
dictionary and then have a smaller dictionary that override specific
entries.

=item B<-E<gt>vendor_num($vendorname)>

Return the vendor number for the given vendor name.

=item B<-E<gt>attr_num($attrname)>

Returns the number of the named attribute.

=item B<-E<gt>attr_type($attrname)>

Returns the type (I<string>, I<integer>, I<ipaddr>, or I<time>) of the
named attribute.

=item B<-E<gt>attr_name($attrnum)>

Returns the name of the attribute with the given number.

=item B<-E<gt>attr_numtype($attrnum)>

Returns the type of the attribute with the given number.

=item B<-E<gt>attr_has_val($attrnum)>

Returns a true or false value, depending on whether or not the numbered
attribute has any known value constants.

=item B<-E<gt>val_has_name($attrnum)>

Alternate (bad) name for I<attr_has_val>.

=item B<-E<gt>val_num($attrnum, $valname)>

Returns the number of the named value for the attribute number supplied.

=item B<-E<gt>val_name($attrnum, $valnumber)>

Returns the name of the numbered value for the attribute number supplied.

=back

There is an equivalent family of accessor methods for Vendor-Specific
attributes and its values. Those methods are identical to their standard
attributes counterparts with two exceptions. Their names have a
I<vsa> prepended to the accessor name and the first argument to each one
is the vendor code on which they apply.

=head1 CAVEATS

This module is mostly for the internal use of Net::Radius::Packet, and
may otherwise cause insanity and/or blindness if studied.

=head1 AUTHOR

Christopher Masto <chris@netmonger.net>, 
Luis E. Muñoz <luismunoz@cpan.org> contributed the VSA code.

=head1 SEE ALSO

Net::Radius::Packet

=cut
